### PR TITLE
Call _isLoopback as a last resort

### PR DESCRIPTION
**Pull request by bleroux**
Call _isLoopback as a last resort

Several developpers complain about seeing ```Invalid argument(s): ArgumentError: Invalid internet address``` for all Http requests (they see this in debug mode and with 'All exceptions' checked).

Main related issue : https://github.com/flutter/flutter/issues/83792 (mixes several issues, some comments might be related to using ```http://``` without allowing it on Android and iOS)
As This is now in Stable (Flutter 2.5), the issue will be reported more and more (example : https://github.com/flutter/flutter/issues/89915)

This could be tracked down to ```_isLoopback``` function relying on a try/on for ```ArgumentError``` : https://github.com/flutter/engine/blob/master/lib/ui/hooks.dart#L238

This PR aims to reduce the annoyance :
- **Before this PR** : ```_isLoopback``` is called for any Http Requests, if the host is not an IP numeric address, ```ArgumentError``` is thrown by ```_InternetAdress.fromString``` and catch by ```_isLoopback```. As using numeric address in Http URIs is not common, it throws almost always. 

- **After this PR** : ```_isLoopback``` is called only when other checks are done, so it won't be call for common Url such as ```https://``` or on platform allowing ```http://``` calls.

@aam and @zanderso , could you review this PR ? And help me to add a test as It's not easy in this particular case ?

_isLoopback was added by this commit : https://github.com/flutter/engine/commit/d46b5b367bb822ccc48f3b0f43caf3d977a13b0c#diff-c105d828b82bd42643a76afa9e91867d72c3d28d6ddd4423392c1f2f4de0ea51

### CODE REVIEW

**Comment by aam** [comment]
Is the issue this PR is addressing actually with AndroidStudio/IntelliJ/VisualCode debugger breaking on this exception as if it is unhandled exception?

**Comment by bleroux** [comment]
> Is the issue this PR is addressing actually with AndroidStudio/IntelliJ/VisualCode debugger breaking on this exception as if it is unhandled exception?

@aam Yes and no :smile:  Yes it is related to debugging experience (IntelliJ/VSCode, but also Flutter DevTools). And yes it is related to IDE breaking on this when the user has asked it ('break on all exceptions'), but this is expected that IDE breaks there when 'All exceptions' is checked.

So what this PR is addressing, is the fact that it happens for all Http request (```_isLoopback``` is calling ```InternetAdress.fromString``` with the host part of the URI ('www.google.com' for instance), despite ```InternetAddress.fromString``` parameter is documented to be a 'numeric address'. So the exception is thrown and catched in almost all Http calls. By moving the _isLoopback call it will be dpne only when really needed.

It will improve debugging experience and also be a little performance improvement (as calling a loopback URL is not so common).

File: lib/ui/hooks.dart [hunk 1]
```diff
@@ -263,6 +260,9 @@ void Function(Uri) _getHttpConnectionHookClosure(bool mayInsecurelyConnectToAllD
         // or if connection is to `https`, allow the connection.
         return;
       }
+      if (_isLoopback(uri.host)) {
```
Comment (aam) [inline lib/ui/hooks.dart:263]: Perhaps update the comment at line 257 since it won't throw if uri is a loopback uri

**Comment by aam** [comment]
> It will improve debugging experience and also be a little performance improvement (as calling a loopback URL is not so common).

Cool, makes sense. Thanks.

**Review by aam — APPROVED** [review]

**Comment by zanderso** [comment]
@aam I believe you had some tests in the original PR for this calback. Is there any way to add to or modify those tests for this change?

**Comment by aam** [comment]
> @aam I believe you had some tests in the original PR for this callback. Is there any way to add to or modify those tests for this change?

Not sure whether test can detect 'breaking on all exceptions' behavior that is visible in the debugger, what this PR addresses.

**Comment by bleroux** [comment]
> > @aam I believe you had some tests in the original PR for this callback. Is there any way to add to or modify those tests for this change?
> 
> Not sure whether test can detect 'breaking on all exceptions' behavior that is visible in the debugger, what this PR addresses.

I wonder too, if 'breaking on all exceptions' could be tested, it would means running in Debug mode with the right flag (if it exists) or by attaching a Debug session.

Another way to test this, would be to detect if ```_isLoopback``` function was called, but I don't know if there is an API for this in Dart tests API (something similar to mockito for instance). Going this way, we could write a test to check that ```_isLoopback``` is not called for URI such as 'https://www.google.com'.

**Comment by bleroux** [comment]
> @aam I believe you had some tests in the original PR for this callback. Is there any way to add to or modify those tests for this change?

@aam @zanderso 
I'm making good progress to write a test (one using VMService). The test is working (or failing :smile: ) as expected on my computer. My problem is that to get 'ArgumentError' thrown, the uri should not be one of those : 'locahost' or an IP address. On my computer I can try with 'http://www.google.com' or 'https://www.flutter.dev/' but what is the policy to have such URLs on test infra ?

Host ```Platform.hostname``` is also a solution for this test, is it allowed on infra ?

**Comment by aam** [comment]
> I'm making good progress to write a test (one using VMService).

Using VMService sounds complicated.
If you want to confirm that zoneOverride check was executed even for loopback hostnames, in your test you can put an instance of your own class into #flutter.io.allow_http: ` zoneValues: <dynamic, dynamic>{#flutter.io.allow_http: myBool(false)}` that would have `bool operator==(Object o)` overridden so it keeps track of whether it was invoked or not.

**Comment by zanderso** [comment]
+1 to @aam's clever suggestion.

**Comment by bleroux** [comment]
> Using VMService sounds complicated.

Yes, it is ! As I finalized my test, I will push it, just that you see how complicated it is :-) (And also to know if it passes on infra).

> If you want to confirm that zoneOverride check was executed even for loopback hostnames, in your test you can put an instance of your own class into #flutter.io.allow_http: ` zoneValues: <dynamic, dynamic>{#flutter.io.allow_http: myBool(false)}` that would have `bool operator==(Object o)` overridden so it keeps track of whether it was invoked or not.

That's really interesting, I will replace my test with this, as it is way simpler and fully coherent with the title I gave to this PR.

**Review by aam — APPROVED** [review]
thank you!

**Review by zanderso — APPROVED** [review]

**Comment by bleroux** [comment]
@zanderso 
With this PR, the issue is fixed on master.  I don't know the policy, do you think this fix could be included in Flutter 2.5.2 or 2.5.3 ?
I ask this, because it seems to frustrate many developpers who relies on stable channel (https://github.com/flutter/flutter/issues/83792)

**Comment by zanderso** [comment]
@bleroux Thanks for raising this. I added the cp label to the issue to start the discussion with the release team.

**Comment by cngeru** [comment]
@zanderso has it been added to the stable channel ?

**Pull request merged**